Initial implementation

simu opened the pull request:
## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the documentation.
- [x] Update the ./CHANGELOG.md.

simu committed:
Initial open-source commit for nfs-client-provisioner component

simu committed:
Update documentation

simu committed:
Update CHANGELOG

simu committed:
Make mountOptions configurable

corvus-ch commented on CHANGELOG.md:
```diff
@@ -4,9 +4,9 @@ All notable changes to this project will be documented in this file.
-- …
+- Initial implementation ([#1])
 
-[Unreleased]: https://github.com/projectsyn/component-nfs-client-provisioner/compare/v0.1.0...HEAD
```
Please keep the link to unreleased. Enter the commit id of the initial commit instead of `v0.1.0`.

simu replied:
seems like unnecessary hassle to have to manually fix this every time, should probably be fixed in the template

corvus-ch commented on class/defaults.yml:
```diff
@@ -1,3 +1,11 @@
 parameters:
   nfs_client_provisioner:
-    namespace: syn-nfs-client-provisioner
+    namespace: syn-storage
```
Not sure if this a good idea. I guess we somehow miss a concept for namespace naming.

simu replied:
reverted to default naming, will need to be rolled out with care on clusters that already use the component

corvus-ch commented on component/main.jsonnet:
```diff
@@ -7,4 +7,5 @@ local params = inv.parameters.nfs_client_provisioner;
 
 // Define outputs below
 {
+  '00_namespace': kube.Namespace(inv.parameters.nfs_client_provisioner.namespace),
```
```suggestion
  '00_namespace': kube.Namespace(params.namespace),
```

srueg replied:
```suggestion
  '00_namespace': kube.Namespace(params.namespace),
```

corvus-ch commented on class/defaults.yml:
```diff
@@ -1,3 +1,11 @@
+    path: /srv/nfs/export
+    mountOptions: null
+    storageClass:
+      defaultClass: true
```
This is dangerous. We have https://github.com/projectsyn/component-storageclass/blob/master/lib/storageclass.libsonnet to ensure there is only one StorageClass with the default annotation.

At least make it default to `false`.

```suggestion
      defaultClass: false
```

simu replied:
removed storageclass config completely as requested by @srueg

srueg commented on class/nfs-client-provisioner.yml:
```diff
@@ -1,11 +1,32 @@
 parameters:
   kapitan:
+    dependencies:
+      - type: https
```
We can use the native helm dependency type

```suggestion
      - type: helm
```

srueg commented on class/defaults.yml:
```diff
@@ -1,3 +1,11 @@
+    server: nfs.local
+    path: /srv/nfs/export
+    mountOptions: null
+    storageClass:
```
Please use the [component-storageclass](github.com/projectsyn/component-storage-class/) to manage storageclasses (and set `create: false` for the chart)

simu replied:
that's gonna get really confusing for users, but ok

srueg replied:
You can have a look at https://github.com/projectsyn/component-csi-cloudscale/blob/master/component/main.jsonnet

corvus-ch commented on docs/modules/ROOT/pages/index.adoc:
````diff
@@ -2,4 +2,22 @@
+
+RHEL / CentOS
+
+```
````
That is markdown, not Asciidoc. Instead of code blocks, I would do this with a definition list.
```
RHEL / CentOS::
  nfs-utils

Ubuntu / Debian::
   nfs-common
```

srueg commented on CHANGELOG.md:
```diff
@@ -4,9 +4,9 @@ All notable changes to this project will be documented in this file.
-- …
+- Initial implementation ([#1])
 
-[Unreleased]: https://github.com/projectsyn/component-nfs-client-provisioner/compare/v0.1.0...HEAD
```
Why are you removing this link? It should point to https://github.com/projectsyn/component-nfs-client-provisioner/compare/1dee7be0a89165228756f70a106e8d8f62cbf5eb...HEAD

srueg commented on component/app.jsonnet:
```diff
@@ -1,4 +1,5 @@
 local kap = import 'lib/kapitan.libjsonnet';
+local kube = import 'lib/kube.libjsonnet';
```
Why importing the kube lib here?

simu committed:
Use component storageclass to create storage class for nfs-client-provisioner

srueg commented on component/main.jsonnet:
```diff
@@ -7,4 +7,5 @@ local params = inv.parameters.nfs_client_provisioner;
 
 // Define outputs below
 {
+  '00_namespace': kube.Namespace(params.namespace),
```
And now I would add the StorageClass here, as in https://github.com/projectsyn/component-csi-cloudscale/blob/master/component/main.jsonnet

simu replied:
Please improve the documentation for component-storageclass, this is highly non-obvious at the moment

corvus-ch commented on component/main.jsonnet:
```diff
@@ -1,10 +1,23 @@
 // The hiera parameters for the component
 local params = inv.parameters.nfs_client_provisioner;
 
+local storageclass = sc.storageClass(params.storageClass.name) {
```
In future, we might have the need to configure more than one storage class (different behavior on delete and or mount options etc.). I would model this as a dictionary containing one or several storage classes. The name would become the a key in the dictionary.

In addition, make this a transparent pass through Instead of hard wiring those fields. Layer different levels of precedence on top of each other (result of the StorageClass function, component wide defaults, values configured per StorageClass). Comment assumes the ability to configure multiple StorageClasses.

simu replied:
This is now implemented, and the component documentation has been updated.

simu committed:
Refactor component to allow configuring multiple storage classes

This change allows users of the component to provide multiple storage
classes with different configurations (e.g. mount options, or archiving
behavior).
